[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor

Le-Zheng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116004
* __->__ #114854
* #114853

**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Le-Zheng committed:
[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -134,6 +134,27 @@ def _may_generate_pattern_with_dtype_convert(pattern, dtype=Arg(), dtype_convert
 )
 
+
+def get_qlinear(users=1):
```
Suggest to use `get_qlinear` for where using `qlinear_pt2e_pattern`, since they can share the code.

Le-Zheng replied:
Done.

jgong5 replied:
`qlinear_pt2e_pattern` is better?

Le-Zheng replied:
Thanks for the comment! I've changed to `qlinear_pt2e_pattern`.

leslie-fang-intel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -569,6 +592,14 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
+                _gelu_fusion_1(get_qlinear(2)),
+                dtype=original_pattern_output_dtype,
+            ),
+            UnaryAttr("gelu", [], ""): generate_pattern_with_output_quant(
```
I believe we should set the `algorithm_attr` for `GeLU` here to distinguish 2 different algorithms.

Le-Zheng replied:
Thank you for your feedback! I've made the requested changes.

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_unary


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -496,6 +517,8 @@ def qconv_binary(match: Match, *args, **kwargs):
 
 
 def _register_quantization_unary_fusion():
+    from .mkldnn_fusion import _gelu_fusion_1, _gelu_fusion_2
```
Thank you for your feedback! I use `_gelu_fusion_erf ` and `_gelu_fusion_tanh` instead.

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -1115,6 +1115,49 @@ def test_qlinear_relu_int8_mixed_bf16(self):
+            self._test_common(
+                mod,
+                (v,),
+                check_autocast=False,
```
Thank you for your feedback! I utilized `int8_mixed_bf16=False` as default checking.

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -1115,6 +1115,49 @@ def test_qlinear_relu_int8_mixed_bf16(self):
+    @skipIfNoDynamoSupport
+    @skipIfNoONEDNN
+    @skipIfRocm
+    def test_qlinear_gelu_cpu(self, int8_mixed_bf16=False):
```
Can we extend and re-use `_qlinear_unary_cpu_test_helper`?

Le-Zheng replied:
Thanks for the comment and changed accordingly.

leslie-fang-intel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -562,11 +571,19 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
+                generate_pattern_with_unary(get_qlinear, aten.relu.default),
+                dtype=original_pattern_output_dtype,
+            ),
+            UnaryAttr("gelu", [], "none"): generate_pattern_with_output_quant(
```
As we synced offline, please follow up the int8-mixed-bf16 cases.

Le-Zheng replied:
create a new pull request to support int8-mixed-bf16 https://github.com/pytorch/pytorch/pull/116004

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -1062,22 +1062,32 @@ def test_qlinear_int8_mixed_bf16(self):
+                    self.unary_fn = copy.deepcopy(unary_op)
                 self.linear2 = torch.nn.Linear(4, 4, use_bias)
-                self.unary_fn2 = torch.nn.ReLU()
+                if unary_op == torch.nn.GELU():
```
2 different GELU instances should not be same always.

Le-Zheng replied:
Thanks for the comment and changed accordingly.

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -1062,22 +1062,32 @@ def test_qlinear_int8_mixed_bf16(self):
                 self.linear = torch.nn.Linear(4, 4, use_bias)
-                self.unary_fn = torch.nn.ReLU()
+                if postop == torch.nn.GELU:
+                    self.unary_fn = postop(approximate=post_op_algo)
```
You are right! Thanks for the comment and changed accordingly.

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion for gelu inside inductor"


**Summary**
Enable QLinear Unary pattern for gelu with int8 

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_cpu


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
